packages/dds/task-manager/src/taskManager.ts

```diff
@@ -273,34 +282,47 @@ export class TaskManager extends SharedObject<ITaskManagerEvents> implements ITa
         });
 
         this.queueWatcher.on("queueChange", (taskId: string, oldLockHolder: string, newLockHolder: string) => {
+            // If oldLockHolder is placeholderClientId we need to emit the task was lost during the attach process
+            if (oldLockHolder === placeholderClientId) {
+                this.emit("lost", taskId);
+                return;
+            }
+
             // Exit early if we are still catching up on reconnect -- we can't be the leader yet anyway.
-            if (this.runtime.clientId === undefined) {
+            if (this.clientId === undefined) {
                 return;
             }
 
-            if (oldLockHolder !== this.runtime.clientId && newLockHolder === this.runtime.clientId) {
+            if (oldLockHolder !== this.clientId && newLockHolder === this.clientId) {
                 this.emit("assigned", taskId);
-            } else if (oldLockHolder === this.runtime.clientId && newLockHolder !== this.runtime.clientId) {
+            } else if (oldLockHolder === this.clientId && newLockHolder !== this.clientId) {
                 this.emit("lost", taskId);
             }
         });
 
         this.connectionWatcher.on("disconnect", () => {
-            assert(this.runtime.clientId !== undefined, 0x1d3 /* "Missing client id on disconnect" */);
+            assert(this.clientId !== undefined, 0x1d3 /* "Missing client id on disconnect" */);
 
             // We don't modify the taskQueues on disconnect (they still reflect the latest known consensus state).
             // After reconnect these will get cleaned up by observing the clientLeaves.
             // However we do need to recognize that we lost the lock if we had it.  Calls to .queued() and
             // .assigned() are also connection-state-aware to be consistent.
             for (const [taskId, clientQueue] of this.taskQueues.entries()) {
-                if (clientQueue[0] === this.runtime.clientId) {
+                if (clientQueue[0] === this.clientId) {
                     this.emit("lost", taskId);
                 }
             }
 
             // All of our outstanding ops will be for the old clientId even if they get ack'd
             this.latestPendingOps.clear();
         });
+
+        runtime.once("attaching", () => {
+            // Once we start attaching we need to release all tasks that were assigned to the placeholder (detached)
+            // client. We do this because we need to re-volunteer with our actual clientId, and there could be a race
+            // to regain the task assignment.
+            this.removeClientFromAllQueues(placeholderClientId);
```

Comment:
> Here I'm wondering if the move should be: instead of always ejecting the placeholder client, instead rewrite the id as the current `this.runtime.clientId` if available, but only eject if the `this.runtime.clientId` is not available?

I think the idea was that we should eject and require the client to re-volunteer because it's possible another client tries to volunteer at the same time we are re-writing the clientId. Do you think this is a concern?

Reply:
I haven't looked in detail, but is the "attaching" event fired before the attach summary is generated or after?  If before, then we have a chance to sneak in and write the client ID.  If after, then yes that concern is def. valid.

Ideally we'd get a chance to do the fixup before the attach summary though.

Reply:
From a quick test from the clicker example, it looks like we summarize before `attaching` is fired

Reply:
Ok maybe one more idea, we could potentially put the logic in summarizeCore itself (if clientId is placeholder, either rewrite to the real clientId if available or else eject the client from the queue)?  Not sure how weird it is to have the logic there, I don't love the side-effect aspect but it's nicer re: avoiding unnecessary disruptions to running tasks.  WDYT?

Reply:
Hey @ChumpChief, one more observation while testing with clicker is that there can be a significant delay between attaching and being assigned a clientId. Although we probably can't rewrite the placeholder most of the time, I think there is still merit to moving the logic before the summary to prevent the placeholder clientId from being in the summary.